### PR TITLE
Update unimod_mapper.py

### PR DESCRIPTION
**Pull request by tobymathieson**
This is a work around for user-defined xml files that have no unimod record ID. Happy to see alternative approaches / suggestions

### CODE REVIEW

**Comment by StSchulze** [comment]
Hi Toby,
thanks for the pull request. In principle that looks fine to, but just to clarify what the use case would be:
this would allow people to write their own unimod xml files (without unimod IDs included) and then use these mods as "normal" unimods in the modifications parameter list?
Wouldn't it be easier to just define those mods in the modifications uparam?
Like this (from docu/description):
> Userdefined modifications need to have a unique name instead of
> the unimod name the chemical composition needs to be given as a
> Hill notation on the fifth position in the string
> 
> Example:
>             ['S,opt,any,New_mod,C2H5N1O3' ]',

Those mods get added into the user-defined unimod xml which can then also be sent around for others to use, etc.

**Comment by tobymathieson** [comment]
Hi Stefan,


Thanks for the feedback and, yes, that is the use cases. And yes your
suggestion was  was my original idea. But then I saw it was possible to
have a user-defined xml, I copied my mascot unimod.xml file over. That
way, *any
and all* modifications added by all users of our system would be included
and I don't have to copy them individually. I figured that's really what a
user-defined unimod xml is. So I blindly added the mascot-derived unimod
xml file and it didn't work. It didn't work because the user-defined.xml
file needs to have a unimod id, which of course cannot exist for
modifications using an in-house compound that binds to Cysteines or a
company standard way of calling SILAC modifications that is more
descriptive than the cryptic 13(L) notation in Unimod, for example.

would be good to hear what you think.

best
Toby

**Comment by StSchulze** [comment]
I see, that makes perfect sense.
The only that I would change is to use "u123" for the user-defined unimod id instead of a number 10001+, since that would keep it consistent with our way of numbering user-defined mods so far and it would prevent overlap with actual unimod IDs (once they get above 10000).

In the future, the requirement for that unimod ID could probably also be removed fairly easily, but I would have to look closer into that.

File: ursgal/unimod_mapper.py [hunk 1]
```diff
@@ -55,6 +55,7 @@ def _parseXML(self, xmlFile = None):
         else:
             xmlFiles = [xmlFile]
         data_list = []
+        extra_unimod_id = 10001
```
Comment (StSchulze) [inline ursgal/unimod_mapper.py:58]: extra_unimod_id = 1
(just initiating a counter here)

File: ursgal/unimod_mapper.py [hunk 2]
```diff
@@ -65,8 +66,13 @@ def _parseXML(self, xmlFile = None):
                 for event, element in unimodXML:
                     if event == b'start':
                         if element.tag.endswith('}mod'):
+                            try:
+                                unimod_id = element.attrib['record_id']
+                            except KeyError:
+                                unimod_id = extra_unimod_id
```
Comment (StSchulze) [inline ursgal/unimod_mapper.py:72]: unimod_id = 'u{0}'.format(extra_unimod_id)

the u prevents any overlap with unimod IDs and also allows us to easily distinguish unimod IDs from user-defined mods.

**Pull request merged**